server/src/main/java/org/elasticsearch/common/settings/Setting.java

```diff
@@ -486,7 +486,11 @@ private T get(Settings settings, boolean validate) {
      * @param defaultSettings the default settings object to diff against
      */
     public void diff(Settings.Builder builder, Settings source, Settings defaultSettings) {
-        if (exists(source) == false) {
+        if (exists(source) == false && exists(defaultSettings) == false) {
```

Comment:
I think this might read better if the outer "doesn't exist in source" stays the same and then have two inner conditions for "if (exists in defaults) X else Y"